JSON operator/function support WIP

mikeplus64 opened the pull request:
Fixes #35 @ChShersh 

WIP

Tricky bits:
1. many functions accept either json or jsonb
2. many functions also overloaded on operand types
3. nullity of many results depends on nullity of either argument; which does not need to be the same

Unimplemented: most of table 9.45 and 9.46 at https://www.postgresql.org/docs/10/static/functions-json.html

echatav commented:
As for (3), you can always embed a `'NotNull` expression as a `'Null` expression using `notNull` so I'd rather avoid defining operations that are so polymorphic they work with different nullities, i.e. your `Null2` type family. If we were to do that consistently then all operators, including arithmetic operators would have to be typed this way. I think it's confusing and breaks some type classes. I also am ok with null safety in Squeal being stronger than what's provided in SQL. A lot of the other type families you define may be unnecessary, `In` should suffice. Other than that, looks good but please put type families in the `Schema` module as that's where I've tried to put all the type-level stuff.

echatav commented:
Oh, and thanks so much for contributing! It's much appreciated.

mikeplus64 commented:
No problem. Squeal is a great library. I think its only real weakness is boilerplate, but this is fairly easy to workaround with magical `generics-sop` stuff or type-reflecting classes...

>A lot of the other type families you define may be unnecessary, In should suffice.
This is true, but I think they're worth it. You get, IMO, *much* more informative type errors from this approach. I think the small boilerplate is worth it for this. To be honest, I'd really like for this approach to be adopted throughout the rest of the library, but I don't really have time to implement this. Error messages are a weak spot to most complex type-trickery Haskell stuff, but it's one that's helped a great deal with appropriate use of `GHC.TypeLits.TypeError`.

I've moved the type families, implemented the rest of table 9.45, and simplified the "nullity" stuff.

mikeplus64 commented:
There is also the problem of functions like

`json_each`
`json_object_keys`

Which return a `setof` something. I.e., they can act like tables. Though they can *also* be used as expressions. See https://www.postgresql.org/docs/10/static/functions-json.html

Maybe there needs to be, added to `PGType`, `PGsetof [(Symbol, PGType)]`? 

Then a function 
```haskell
type NotNulls :: [(Symbol, PGType)] -> [(Symbol, NullityType)] -- probably exists already somewhere?
setof 
  :: Aliased (Expression schema relations grouping) (alias ::: nullity ('PGsetof ty))
  -> FromClause schema params (NotNulls ty)
```

mikeplus64 commented:
*and* then a function to cast a `setof` as an `PGcomposite`...

Or is this just something that is possible with *any* composite-value-returning-expression? I have no idea. It is all fairly magical to me still. But if so maybe all that's needed is

```haskell
rows
  :: Aliased (Expression schema relations grouping) (alias ::: nullity ('PGcomposite ty))
  -> FromClause schema params (NotNulls ty)
```

echatav commented:
Oooh, that's a smart idea, the thing is once we change composites to have nullities for fields then they are interchangeable with relations, that is:

```haskell
PGcomposite :: RelationType -> PGType
```

I had been thinking more about how that means that we can generate a composite type from any table or view, but it also means exactly what you said, basically the opposite direction; generate a table expression from a single expression of composite type.

echatav commented:
Looking at the docs a little more closely, it appears that `setof` types are sets in the sense of rows of a table. These aren't proper types, but only from clauses so better perhaps to bypass writing these as functions targeting expressions and just target from clauses? That or don't bother with those functions. Exhaustive support for all the tables (9.43-9.46) is unneccessary, so long as Squeal obtains feature parity with other libs :-)

mikeplus64 commented:
This now implements everything on https://www.postgresql.org/docs/10/static/functions-json.html

Should be ready for merge/review. Stuff left is actually testing it, and `json_agg`/`jsonb_agg` but I figure those might be better put into a patch providing `array_agg` and company.

mikeplus64 commented on squeal-postgresql/src/Squeal/PostgreSQL/Query.hs:
```diff
@@ -623,6 +640,185 @@ offset
+  (renderAliasedAs renderExpression aliasedExpr)
+
+-- | Expands the outermost JSON object into a set of key/value pairs.
+jsonEach
```
Given a JSON object like `{ "key": value }` it produces a table `(key string, value jsonb)`;

```
psql=# select t.* from json_each('{"x":3,"y":{"z":12}}') as t;
 key |  value   
-----+----------
 x   | 3
 y   | {"z":12}

```

echatav replied:
Ok, I take this to mean that `json_each` can only be applied to single values, that is it cannot refer to any columns in any tables. In this case it's closed with respect to the `relations` type parameter. There are two ways of representing this, with a frankentype or the empty list.

```haskell
(forall relations. Expression schema relations 'Ungrouped params)
```
or
```haskell
Expression schema '[] 'Ungrouped params
```

I'd pick the latter.

echatav replied:
I don't think Squeal can support such queries because `FromClause`s don't have a relations parameter which would be necessary for them to reference columns in another relation previously defined. Look at it this way, you have right now a type signature:

```Haskell
jsonEach
  :: Aliased (Expression schema relations 'Ungrouped params) '(tab, nullity 'PGjson)
  -> FromClause schema params (PGjson_each tab)
```

The `relations` parameter gets lost. It might be a later feature to add one into from clauses, but it doesn't seem that useful.

Also thanks for adding json literals! That's super useful I think.

mikeplus64 commented:
Anything left before merge?

echatav commented:
I don't think there's anything left. I will merge this when there are enough features to warrant a 0.4 release. Thanks!

echatav commented:
I decided to do a release just with these new JSON functions because there's too many other things that might otherwise be included in a 0.4 release. I still don't quite understand version numbering policy but since this PR doesn't change Squeal's type system I'll make it a minor version update.
